Add Jörmungandr Servant Api Type

Anviking opened the pull request:
# Issue Number

#219 


# Overview

- [x] I added a `FromBinary` class in `Cardano.Wallet.Binary.Jörmungandr`
- [x] I added `Cardano.Wallet.Network.Jörmungandr.Api` where the `OctetStream` content type relies on the `FromBinary` class.

# Comments

- I am not sure about the organisation of logic here, e.g `FromBinary`, orphans, etc

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

KtorZ commented on src/Cardano/Wallet/WalletLayer/Kernel/Wallets.hs:
```diff
@@ -281,10 +286,11 @@ addressPoolGapByRootId rootId db = do
        | not . null $ errors   -> Left $ GetEosWalletErrorWrongAccounts anId
        | length (nub gaps) > 1 -> Left $ GetEosWalletErrorGapsDiffer anId
-       | otherwise             -> let [gap] = gaps in Right gap
+       | otherwise             -> let gap:_ = gaps in Right gap
```
Didn't quite spot that one in the review :upside_down_face: ... But, gaps are only `nubs` in the branch before, so here, they are still a full list and `[gap]` doesn't pattern match well with more than one account.

KtorZ commented:
@denisshevchenko Did you got time to look into this ?

Anviking committed:
Add Cardano.Wallet.Jormungandr.Network.Api

- Servant type in the .Api module, with Plain and Binary content types.
- I added a FromBinary class in the binary module to play nicely with the Binary content-type

Anviking commented on lib/jormungandr/src/Cardano/Wallet/Binary/Jormungandr.hs:
```diff
@@ -67,13 +72,15 @@ import Data.Quantity
     , chainLength :: Word32
     , contentHash :: Hash "content"
-    , parentHeaderHash :: Hash "parentHeader"
+    , parentHeaderHash :: Hash "BlockHeader"
```
To match with the one in `Cardano.Primitive.Types (BlockHeader)`

Anviking commented on lib/jormungandr/src/Cardano/Wallet/Binary/Jormungandr.hs:
```diff
@@ -364,3 +371,33 @@ whileM cond next = go
+instance FromBinary Block where
+    get = getBlock
+
+instance FromBinary W.Block where
```
The `Block` type in this module contain more information than the one in `Cardano.Wallet.Primitive.Types` (`W.Block`).

Anviking commented on lib/jormungandr/src/Cardano/Wallet/Network/Jormungandr/Api.hs:
```diff
@@ -0,0 +1,116 @@
+-- TODO: Replace SignedTx with something real
+data SignedTx
+
+newtype BlockId = BlockId (Hash "block")
```
Could also have `ApiT`, but the jörmungandr api use the term "BlockId" which this newtype can capture.

Anviking commented on lib/jormungandr/src/Cardano/Wallet/Network/Jormungandr/Api.hs:
```diff
@@ -0,0 +1,116 @@
+        BlockId . Hash <$> convertFromBase Base16 (BL.toStrict bs)
+
+-- Orphan instance
+instance FromBinary a => MimeUnrender OctetStream a where
```
Orphan instance can be avoided by defining `OctetStream` ourselves in this module. I did that at first.

KtorZ commented on lib/jormungandr/src/Cardano/Wallet/Network/Jormungandr/Api.hs:
```diff
@@ -0,0 +1,116 @@
+    :> "block"
+    :> Capture "blockId" BlockId
+    :> QueryParam "count" Int
+    :> Get '[OctetStream] [BlockId]
```
I am not sure `OctetStream` is the best fit here. We actually know the format of the API we're talking to. So it's not just plain bytes. I'd suggest to define some custom format like `Jormungandr` or `BetterCBOR` or any name that fits to describe the format, and makes it clear that we're still expecting things in a particular format (whereas OctetStream conveys the idea of a stream of bytes without any particular structure). 

This will also prevent the orphan :+1:

Anviking replied:
👍
